waku/v2/protocol/waku_enr/sharding.nim

```diff
@@ -0,0 +1,222 @@
+  if topic.kind != NsPubsubTopicKind.StaticSharding:
+    return false
+
+  rs.contains(topic.cluster, topic.shard)
+
+func contains*(rs: RelayShards, topic: PubsubTopic|string): bool =
+  let parseRes = NsPubsubTopic.parse(topic)
+  if parseRes.isErr():
+    return false
+
+  rs.contains(parseRes.value)
+
+
+# ENR builder extension
+
+func toIndicesList(rs: RelayShards): seq[byte] =
+  if rs.indices.len > 255:
+    raise newException(Defect, "indices list too long")
+
+  var res: seq[byte]
+  res.add(rs.cluster.toBytesBE())
+
+  res.add(rs.indices.len.uint8)
+  for index in rs.indices:
+    res.add(index.toBytesBE())
+
+  res
+
+func fromIndicesList(buf: seq[byte]): Result[RelayShards, string] =
+  if buf.len < 3:
+    return err("insufficient data")
+
+  let cluster = uint16.fromBytesBE(buf[0..1])
+  let length = int(buf[2])
+
+  if buf.len != 3 + 2 * length:
+    return err("invalid data")
+
+  var indices: seq[uint16]
+  for i in 0..<length:
+    indices.add(uint16.fromBytesBE(buf[3 + 2*i ..< 5 + 2*i]))
+
+  ok(RelayShards(cluster: cluster, indices: indices))
+
+func toBitVector(rs: RelayShards): seq[byte] =
+  ## The value is comprised of a two-byte shard cluster index in network byte
+  ## order concatenated with a 128-byte wide bit vector. The bit vector
+  ## indicates which shards of the respective shard cluster the node is part
+  ## of. The right-most bit in the bit vector represents shard 0, the left-most
+  ## bit represents shard 1023.
+  var res: seq[byte]
+  res.add(rs.cluster.toBytesBE())
+
+  var vec = newSeq[byte](128)
+  for index in rs.indices:
+    vec[index div 8].setBit(index mod 8)
+
+  res.add(vec)
+
+  res
+
+func fromBitVector(buf: seq[byte]): Result[RelayShards, string] =
+  if buf.len != 130:
+    return err("invalid data")
+
+  let cluster = uint16.fromBytesBE(buf[0..1])
+  var indices: seq[uint16]
+
+  for i in 0u16..<128u16:
+    for j in 0u16..<8u16:
+      if not buf[2 + i].testBit(j):
+        continue
+
+      indices.add(j + 8 * i)
+
+  ok(RelayShards(cluster: cluster, indices: indices))
+
+
+func withWakuRelayShardingIndicesList*(builder: var EnrBuilder, rs: RelayShards) =
+  builder.addFieldPair(ShardingIndicesListEnrField, rs.toIndicesList())
+
+func withWakuRelayShardingBitVector*(builder: var EnrBuilder, rs: RelayShards) =
+  builder.addFieldPair(ShardingBitVectorEnrField, rs.toBitVector())
+
+func withWakuRelaySharding*(builder: var EnrBuilder, rs: RelayShards) =
+  if rs.indices.len >= 64:
+    builder.withWakuRelayShardingBitVector(rs)
+  else:
+    builder.withWakuRelayShardingIndicesList(rs)
+
+
+# ENR record accessors (e.g., Record, TypedRecord, etc.)
+
+func relayShardingIndicesList*(record: TypedRecord): Option[RelayShards] =
+  let field = record.tryGet(ShardingIndicesListEnrField, seq[byte])
+  if field.isNone():
+    return none(RelayShards)
+
+  let indexList = fromIndicesList(field.get())
+  if indexList.isErr():
+    return none(RelayShards)
```

Comment:
In this case we are losing the possible error given by `fromIndicesList`.
I'd try to chain the possible errors back to the caller as much as possible by returning a `Result[Option[..], string]`